### PR TITLE
fix(models): drop stale Codex fallback entries

### PR DESCRIPTION
**Pull request by starship-s**
## Thinking Path

- The `openai-codex` fallback belongs to the ChatGPT/Codex subscription provider. Whether the same model ID remains active through the OpenAI API does not establish that it remains selectable through this provider.
- Account-aware live discovery and the visible local Codex cache are authoritative. The static list is only the degraded fallback.
- Generic Agent-core seeding bypassed that provider-specific boundary and could silently restore retired or entitlement-dependent Codex entries.
- Live-only Codex model IDs still need explicit provider context at runtime. Otherwise, an overlapping configured provider can claim the same bare model ID.

## What Changed

- Removed the nonexistent `gpt-5.5-mini` entry from the static OpenAI, OpenAI API, and OpenAI Codex fallbacks.
- Removed `gpt-5.3-codex`, `gpt-5.2-codex`, `gpt-5.1-codex-max`, `gpt-5.1-codex-mini`, and `codex-mini-latest` from the static OpenAI Codex fallback.
- Kept `gpt-5.3-codex-spark` out of the unconditional fallback because the fallback has no account-tier evidence.
- Excluded `openai-codex` from generic Agent-core seeding. Its existing live/cache path remains responsible for account-specific catalog freshness, including Spark for entitled accounts.
- Qualified Codex selections as `@openai-codex:<model>` before same-provider bare passthrough. This preserves Codex routing when another configured provider advertises the same live-only model ID.
- Updated the focused fallback, seeder, and routing tests. Removed a stale Spark-test skip that triggered even when the intended Codex lookup occurred.

## Why It Matters

- Degraded catalog loading no longer advertises retired or entitlement-dependent subscription choices.
- API model support remains separate from the ChatGPT/Codex subscription fallback.
- Live-discovered Codex models cannot be redirected to an overlapping configured provider.
- The Codex provider group remains explicit through its provider ID, authentication, and routing. It does not need a model ID containing `codex` as a UI sentinel.

## Verification

Focused provider, resolver, seeder, and Codex tests:

`./scripts/test.sh tests/test_provider_mismatch.py tests/test_model_resolver.py tests/test_opencode_providers.py tests/test_4413_seed_provider_models.py tests/test_issue1189_openai_codex_detection.py tests/test_issue1680_codex_spark.py tests/test_issue1807_codex_provider_card_live_models.py -q`

Result: `165 passed`

Routing regression:

- Before the fix, a live-only `gpt-5.3-codex` selection resolved to an overlapping `shared-relay` provider.
- After the fix, the same selection resolves to `openai-codex` and preserves the configured Codex base URL.

Additional checks:

- `python3 -m py_compile` on all changed Python files: passed
- `git diff --check`: passed
- Merge-tree against current `master`: clean

## Risks / Follow-ups

- Live discovery or the visible local Codex cache can still supply Spark and other account-authorized models by design.
- No screenshot is included because this is catalog-data and routing-boundary cleanup with no layout or interaction change.

## Model Used

- OpenAI / GPT-5.6 Sol for investigation, orchestration, implementation follow-up, and verification.
- OpenAI / GPT-5.6 Luna for implementation and independent review.

### CODE REVIEW

File: tests/test_opencode_providers.py [hunk 1]
```diff
@@ -118,14 +120,11 @@ def test_shared_opencode_api_key_detects_zen_and_go(monkeypatch):
...
-    assert "openai-codex" in config._PROVIDER_MODELS
-    ids = [m["id"] for m in config._PROVIDER_MODELS["openai-codex"]]
-    assert "gpt-5.4" in ids, f"gpt-5.4 missing from openai-codex catalog: {ids}"
-    assert "gpt-5.4-mini" in ids, f"gpt-5.4-mini missing from openai-codex catalog: {ids}"
-    assert "gpt-5.3-codex" in ids, f"gpt-5.3-codex missing from openai-codex catalog: {ids}"
-    assert "gpt-5.2-codex" in ids, f"gpt-5.2-codex missing from openai-codex catalog: {ids}"
+def test_openai_codex_static_fallback_includes_spark():
+    """The literal Codex fallback keeps its current Codex-specific model."""
+    source = Path(config.__file__).read_text(encoding="utf-8")
+    codex_block = source.split('    "openai-codex": [', 1)[1].split('    ],\n    "google":', 1)[0]
```
[the remark this answers is not in the document]
Comment (starship-s) [reply tests/test_opencode_providers.py]: Fixed in `d7a567e`: removed the formatting-sensitive source parser and moved the Spark assertion into the existing empty-live-catalog provider test, which exercises the runtime fallback path. Targeted tests pass (`12 passed`).

**Comment by nesquena-hermes** [comment]
## Code gate is clean ✅ — parking for a maintainer catalog-confirmation (not a rejection)

Thanks @starship-s. I ran the full gate (Codex correctness pass + 5-shard suite) on a rebased head, and the **code is sound**:

- ✅ Referential integrity: none of the removed ids (`gpt-5.5-mini`, `gpt-5.3-codex`, `gpt-5.2-codex`, `gpt-5.1-codex-max`, `gpt-5.1-codex-mini`, `codex-mini-latest`) are referenced elsewhere as a default, mapping key, or capability lookup — no dangling refs.
- ✅ No provider default resolves to a removed id.
- ✅ The added `gpt-5.3-codex-spark` is consistent with GPT-5/Codex routing + capability logic, and is already known to the codebase (`api/config.py:6460` references it from #1680).
- ✅ Live account + Agent catalogs remain authoritative and unchanged — this is fallback-only, not a global filter. Agent seeding can still restore Agent-authoritative ids.
- ✅ The 3 test edits reflect the new catalog without concealing a runtime break. 24 targeted tests pass.

The one thing the gate **can't** verify is the PR's premise itself: whether those specific older Codex ids are genuinely retired / no-longer-valid fallback choices. That's an external model-catalog fact, and since the change removes user-facing picker entries on that basis, I'm leaving it for a maintainer to confirm rather than merging it autonomously.

**My recommendation: ship it.** The change is low-risk (fallback-only, live/agent catalogs authoritative, added id validated in-repo), it reduces stale surface, and it's on the experimental channel. The only reason it's not already merged is that confirming "these 5 Codex ids are actually dead" needs a human who knows the current catalog — a 10-second check for a maintainer. If any of the removed ids is still valid, just say so and I'll adjust the fix; otherwise this is good to go.

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
Thanks @starship-s — most of this is correct and verified: `gpt-5.5-mini` is genuinely absent from the current catalog, the other removed Codex IDs (`gpt-5.2-codex`, `gpt-5.1-codex-max`, `gpt-5.1-codex-mini`, `codex-mini-latest`) have passed their shutdown dates, live/cache catalogs remain authoritative, and 17 focused tests pass. Two edges need correcting before this ships, though:

### Must-fix

1. **[SILENT] `api/config.py:1706` — don't remove `gpt-5.3-codex`.** OpenAI still lists `gpt-5.3-codex` as an **active API model** (its docs call it *deprecated*, not *retired*), and the matching Hermes Agent 0.20.2 core still retains it. Dropping it from the `openai-codex` degraded fallback means a degraded/offline installation loses a model users can still legitimately select. **Fix:** restore `{"id": "gpt-5.3-codex", "label": "GPT-5.3 Codex"}`. (ref: https://developers.openai.com/api/docs/models/gpt-5.3-codex)

2. **[CORE] `api/config.py:1710` — don't add `gpt-5.3-codex-spark` to the *unconditional* fallback.** Spark is officially **Pro-only**, but this static fallback carries no account-tier evidence, so a non-Pro Codex user is offered Spark during degraded discovery and then hits an **entitlement failure when they actually send** with it selected. A static fallback can't know entitlement. **Fix:** remove Spark from the unconditional fallback (and from `tests/test_issue1807_codex_provider_card_live_models.py:89`); let Spark surface only from the account-authoritative live/cache catalog where entitlement is known. (ref: https://learn.chatgpt.com/docs/models)

Net: keep the four genuinely-retired removals, **restore `gpt-5.3-codex`**, and **drop the Spark fallback addition**. With those two changes this is a clean catalog-hygiene fix. Re-push and I'll re-gate.

**Comment by starship-s** [comment]
@nesquena-hermes Follow-up pushed in `bfaa08b0`. Addressing both review points:

1. `gpt-5.3-codex` remains removed from the static `openai-codex` fallback. That fallback is for the ChatGPT/Codex subscription provider, so availability through the OpenAI API does not establish availability through this auth/routing path. The provider group is identified by `provider_id == "openai-codex"` and does not need a model ID containing `codex` as a sentinel.

2. Agreed on Spark. `gpt-5.3-codex-spark` is now removed from the unconditional fallback. I also found that `_seed_provider_models_from_core()` would re-add both Spark and `gpt-5.3-codex` at import time, bypassing the same account-entitlement boundary. `openai-codex` is now excluded from generic core seeding; its existing account-aware live/cache discovery remains authoritative and still surfaces Spark for entitled accounts.

I also removed the stale Spark-test skip and added seeder-boundary coverage. Focused tests are `25 passed`, and all 23 exact-head GitHub checks are green. Please re-gate when ready.

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
## Gate at head `bfaa08b0b2dd`: the redesign introduces a Codex routing regression

Thanks @starship-s. The catalog-hygiene intent is right — the four genuinely-retired IDs (`gpt-5.2-codex`, `gpt-5.1-codex-max`, `gpt-5.1-codex-mini`, `codex-mini-latest`) and `gpt-5.5-mini` are correctly gone, and the entitlement concern that motivated the last review is real.

But the approach here diverges from the requested surgical fix: instead of restoring `gpt-5.3-codex` and dropping only Spark from the `openai-codex` static fallback, this **removes all Codex IDs from the static fallback and excludes `openai-codex` from `_seed_provider_models_from_core()`**, delegating the whole catalog to the live/cache path. I gated the redesign on its own merits (adversarial reproduction + full suite, exact head rebased onto current master), and it has one **CORE** regression that blocks it.

### CORE — live-discovered Codex models can be routed to the wrong backend (api/config.py:1948 → ownership set 2686–2726 → `model_with_provider_context()` :3134)

Excluding `openai-codex` from core-seeding removes those model IDs from the **ownership set** (`_provider_models_set`) that `model_with_provider_context()` relies on. When a Codex model then comes from the live catalog, it falls through to the bare-model passthrough at :3134 (`if provider == config_provider: return model`) instead of emitting an explicit `@openai-codex:<model>` routing hint. If an overlapping custom/configured provider's catalog carries the same model ID, the bare model gets **hijacked** — this is the exact `#5511`/`#5909` hijack class the surrounding code already guards against (see the comments at :3124–3131 and :2686–2693).

Reproduced: `gpt-5.3-codex` with `model_provider=openai-codex` resolved to `custom:shared-relay` on this branch, but correctly to `openai-codex` with the master catalog restored. A Codex turn can be sent to the wrong backend.

**Why the surgical fix avoids this:** keeping the Codex IDs in the `openai-codex` static catalog (minus the dead ones + Spark) keeps them in the ownership set, so the routing hint is preserved. That's the lower-risk path the earlier review asked for.

### Two ways forward
1. **Surgical (recommended, matches the prior review):** keep the four retired removals, **restore `gpt-5.3-codex`** to the `openai-codex` static list, **drop only `gpt-5.3-codex-spark`** (Pro-only, entitlement-unknown in a static list), and do NOT exclude `openai-codex` from core-seeding. This gets the hygiene win without touching routing.
2. **Keep the redesign** only if you also preserve Codex routing context: return `@openai-codex:<model>` in `model_with_provider_context()` before the same-provider bare return at :3134 for live-discovered Codex models, and add the overlapping-custom-provider hijack case as a regression test.

The degraded/offline picker behavior itself is safe (with discovery empty it shows the three static models and injects a configured removed default without crashing) — the blocker is purely the routing hijack. Re-push either path and I'll re-gate.


---
_Gate evidence: full suite **14,684 passed / 0 failed** at exact head (clean); the block is the single reproduced routing regression above._
